Workflow changes management

beygorghor opened the pull request:
Allow user to manage workflow changes  (CRUD)
See [Whimsical](https://whimsical.com/web-interface-5JjLtfS6my7DwLqqoaaTTT) for mockups

Changes are containing  mapping between target (a form you have to select) and source (the reference form of a workflow)
By adding a mapping, you'll decide if a field of the target should affect one of the field of the reference form. 

Related JIRA tickets : WC2-113, WC2-114, WC2-129, WC2-130, WC2-132, WC2-122


## Self proof reading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] Did I add translations
- [x] My migrations file are included
- [ ] Are there enough tests

## Changes

- bask-end api to CRUD changes for a specific workflow version
- front-end adaptation to make it possible
- fix search in workflow versions list
- fix deep copy of a workflow version

## How to test

Go to a list of workflow versions  for a beneficiary type, try to search a version by name
Try to make a duplication of a version
Go to the detail of a draft workflow version.
Play with the changes keeping in mind that:
- Source and target should be of the same type
- You cannot have multiple mappings on the same source or target
- target form can be used only once in the changes


## Print screen / video

https://user-images.githubusercontent.com/12494624/214072335-f64a8c23-3ac5-4fba-9748-36af7c282538.mov


## Notes

Test will be done on separate PR's (WC2-132, WC2-133)
Sorry for the huge PR

olethanh commented on hat/assets/js/apps/Iaso/index.js:
```diff
@@ -12,7 +12,13 @@ import { store, history } from './redux/store';
 import { getPlugins, PluginsContext } from './utils';
 
-const queryClient = new QueryClient();
+const queryClient = new QueryClient({
+    defaultOptions: {
+        queries: {
+            refetchOnWindowFocus: false,
+        },
+    },
+});
```
not really part of the table stuff but I'm all for it

olethanh commented on plugins/polio/js/src/components/Dashboard.js:
```diff
@@ -3,11 +3,10 @@ import React, { useCallback, useEffect, useMemo, useState } from 'react';
+    IconButton as IconButtonComponent,
 } from 'bluesquare-components';
-import 'react-table/react-table.css';
+// import 'react-table/react-table.css';
```
remove comment

quang-le reviewed (changes requested):
Nice job, I like the new look of the table (I think the little shadow effect makes it classier than the previous one).
Code review is in the comments. Nothing blocking, except maybe the condition in `useGetPages`

I tested the app, here's what I found:

- You need to click twice for the sorting happen, but only in one direction. It seems the call is not made on the first click:

https://user-images.githubusercontent.com/38907762/130796080-f9a971cf-3122-4508-bc72-5063213f251f.mov

- In Polio and Pages, sorting will cause the page to flicker:

polio

https://user-images.githubusercontent.com/38907762/130796307-c65ed9b3-3cb9-44b7-a590-027eb3f878b2.mov

pages


https://user-images.githubusercontent.com/38907762/130796428-07bbade4-57af-49d2-bd41-4054c6b52048.mov


- History screen can crash because of a null error:


https://user-images.githubusercontent.com/38907762/130796659-3e948596-b2d6-417f-9086-1ada0b23dc92.mov


- In History screen, when sorting while the row is expanded, the expanded row remains the same, while the table sorting order changes:


https://user-images.githubusercontent.com/38907762/130797111-1399c35d-59de-44ad-a291-74ac92bb63a3.mov


-  When sorting Tables with lots of data, clicking on the column header multiple times will create as many calls, and as many refreshes of the Table. The previous version had a loading screen which hid this to the user. I think we should either reintroduce a loading screen, or disabling sorting while the awaiting the request's response:


https://user-images.githubusercontent.com/38907762/130797739-9fa50646-f75a-41f3-9c5b-00a68feaa808.mov


- In the SideBar, the link to Archived (forms) redirects to List:


https://user-images.githubusercontent.com/38907762/130798104-f17e147a-415c-4acb-9f14-b6aa348755a3.mov


- The Algorithms page is broken:

https://user-images.githubusercontent.com/38907762/130798284-427258d7-aded-460c-9c81-7e13796f582d.mov


- When resizing, the cursor doesn't stick to the header, making it difficult to control with large sizes:


https://user-images.githubusercontent.com/38907762/130798584-06b28380-7e1d-438a-98be-d9dd3b2a61be.mov



https://user-images.githubusercontent.com/38907762/130798624-f70e0f7d-fc5a-49e6-897e-df958ead8744.mov


- The cursor changes to text when hovering over a non-sortable table header. Previous behaviour was to use default cursor:


https://user-images.githubusercontent.com/38907762/130799893-96e45c69-2648-443f-8e80-a9424e44a074.mov


- The tooltip for table headers is not styled like other tooltips (eg: action buttons). The texxt could also be changed to something like "Toggle sort order" or "Change sort order":

![Screenshot 2021-08-25 at 12 24 08](https://user-images.githubusercontent.com/38907762/130800207-13721233-00f4-4e4e-9926-9591d9c18a56.png)


- The pagination doesn't have enough room to display page counts over 99:

![Screenshot 2021-08-25 at 13 58 48](https://user-images.githubusercontent.com/38907762/130800377-a271aa6c-a2a6-4cb4-8935-c5a932ff78b9.png)


Sorry, lots of nitpicking because so many parts of the app are impacted. I can't wait to use it though!

Great job!

beygorghor commented on hat/assets/js/test/utils/redux.js:
```diff
@@ -56,7 +58,7 @@ const initialState = {
         <LinkProvider linkComponent={Link}>
-            {renderWithIntl(component)}
+            {renderWithMuiTheme(renderWithIntl(component))}
         </LinkProvider>
```
Yes , create a ticket please

beygorghor commented on hat/assets/js/apps/Iaso/components/filters/chips/ChipsFilterComponent.js:
```diff
@@ -112,7 +112,7 @@ class ChipsFilterComponent extends Component {
                     <InputComponent
-                        keyValue="form_id"
+                        keyValue={`${locationsKey}-not-selected`}
                         onChange={(key, formId) => this.onAdd(formId)}
```
yep

olethanh commented on hat/assets/js/apps/Iaso/domains/completeness/messages.js:
```diff
@@ -33,7 +33,7 @@ const MESSAGES = defineMessages({
         id: 'iaso.label.months.june',
     },
-    july: {
+    jully: {
```
err?

beygorghor committed:
WIP on change modal

beygorghor committed:
split code for version, follow ups and changes

beygorghor committed:
WIP edit mapping

beygorghor committed:
Merge branch 'main' into WC2-114-workflow-changes

beygorghor committed:
mapping table, delete, create, change

beygorghor committed:
disable save if mapping is not valid

beygorghor committed:
UX improvements

beygorghor committed:
Merge branch 'main' into WC2-114-workflow-changes

Commit:
Initial Work on WC2-113 and bugfix deepcopy change and text search

mathvdh committed:
Workflow Change creation works and save, needs to validate the mapping types still

mathvdh committed:
Fix for problem of reference_form on WorkflowVersion and on EntityType

mathvdh committed:
Checking the types of the mapping works now

mathvdh committed:
Deleted docs

mathvdh committed:
Modifications for workflow change add/edit/delete

Commit:
Frontend messages update

mathvdh committed:
Message after delete

mathvdh committed:
FIX merge conflict

Commit:
Workflow Changes OK niveau backend

beygorghor committed:
Merge branch 'main' into WC2-113-edit-changes-api

beygorghor committed:
merge migrations

beygorghor committed:
removing ts ignore and fix typescript errors

beygorghor committed:
fix changes api

- update form changes
- fix validate mappings

beygorghor committed:
validate changes modifications

- only use a form once
- Source and target should be of the same type
- Cannot have multiple mappings on the same source or target

beygorghor committed:
rename possibleFields

beygorghor committed:
Merge branch 'main' into WC2-114-changes-validation

quang-le commented on hat/assets/js/apps/Iaso/domains/app/translations/fr.json:
```diff
@@ -896,17 +896,27 @@
     "iaso.workflows.addWorkflowVersion": "Ajouter une version",
     "iaso.workflows.changes": "Changements",
+    "iaso.workflows.changesInfo": "La source et l'origine doivent avoir le même type",
+    "iaso.workflows.changesInfo2": "Plusieurs correspondances avec la même origine ou la même source est impossible",
```
I think it's correct French, but it sounds funny

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/components/ModalButtons.tsx:
```diff
@@ -1,10 +1,5 @@
-    // @ts-ignore
-    IconButton,
-} from 'bluesquare-components';
+import { useSafeIntl, IconButton } from 'bluesquare-components';
```
😀

quang-le commented on hat/webpack.dev.js:
```diff
@@ -5,7 +5,17 @@ const BundleTracker = require('webpack-bundle-tracker');
+    WEBPACK_HOST = 'localhost';
+}
+
+const WEBPACK_URL = 'http://'+ WEBPACK_HOST +':3000';
```
Not sure I understand the need, but why not

mathvdh replied:
It could be removed honestly ... The need is if you want to access a development version from another computer than the one running the webpack dev server (for example if you want to test on a tablet or iphone on the same local network) .. But OK it could be removed ...

beygorghor replied:
May be add is into a separate PR

mathvdh commented:
@beygorghor  A la limite faudrait pas merger mes tests avec ça que ça passe ?

mathvdh committed:
Modifs .gitignore

mathvdh committed:
Reimport tests from WC2-132 branch

mathvdh committed:
extract BASE_API and cleanup urls

mathvdh committed:
Merging migrations

mathvdh committed:
wf changes 100% coverage

mathvdh committed:
Much better tests

mathvdh committed:
Cleanup calendar file

mathvdh commented:
On y est niveau tests .. Après 100% de coverage ça garantit pas que le code en lui même n'a pas oublié des verifications et qu'il est bon .. 

<img width="829" alt="Screenshot 2023-01-26 at 22 01 31" src="https://user-images.githubusercontent.com/383344/214951368-b7d69834-1c4c-4c4e-8d90-d62fa26f0d6d.png">

mathvdh committed:
Black formatting

mathvdh committed:
Remove prettyprinter was used for dev

mathvdh reviewed (approved):
Fine for me !

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/components/changes/Modal.tsx:
```diff
@@ -0,0 +1,241 @@
+import { PossibleField } from '../../../forms/types/forms';
+import { useGetPossibleFields } from '../../../forms/hooks/useGetPossibleFields';
+
+type Props = {
+    isOpen: boolean;
+    closeDialog: () => void;
+    change?: Change;
+    versionId: string;
+    targetPossibleFields: PossibleField[];
+    referenceForm?: ReferenceForm;
+    changes?: Change[];
+};
```
One day I will make a type in bluesquare-components with the props that come from modal HoC, that would make it clearer when we define `Props`

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/components/changes/Modal.tsx:
```diff
@@ -0,0 +1,241 @@
+                form,
+                mapping: mappingObject,
+            });
+        }
```
Not blocking, but I think it was done in a more efficient way in `useSaveCampaign`:  There's only one hook, but the mutation function make a post or a put depending on the presence (or not) of the id.

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/components/changes/Modal.tsx:
```diff
@@ -0,0 +1,241 @@
+                ),
+            ).length === mappingArray.length,
+        [mappingArray, sourcePossibleFields],
+    );
```
You don't need to memoize a boolean

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/components/changes/Modal.tsx:
```diff
@@ -0,0 +1,241 @@
+            cancelMessage={MESSAGES.cancel}
+            confirmMessage={MESSAGES.confirm}
+            open={isOpen}
+            closeDialog={() => null}
```
Should probably make this one optional in blsq-comp

beygorghor replied:
yes

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/components/changes/Modal.tsx:
```diff
@@ -0,0 +1,241 @@
+            closeDialog={() => null}
+            dataTestId="workflow-change"
+            id="workflow-change"
+            onClose={() => null}
```
idem

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/components/changes/MappingTable.tsx:
```diff
@@ -0,0 +1,186 @@
+
+    const handleUpdate = useCallback(
+        (
+            key: 'target' | 'source',
```
you can use `key: keyof Mapping` for this

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/components/changes/SourceCell.tsx:
```diff
@@ -0,0 +1,81 @@
+    sourceOptions: ChangesOption[];
+    handleUpdate: (
+        // eslint-disable-next-line no-unused-vars
+        key: 'target' | 'source',
```
cf supra: key : keyof Mapping

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/components/changes/TargetCell.tsx:
```diff
@@ -0,0 +1,80 @@
+    sourceOptions: ChangesOption[];
+    handleUpdate: (
+        // eslint-disable-next-line no-unused-vars
+        key: 'target' | 'source',
```
keyof etc.

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/components/changes/TargetCell.tsx:
```diff
@@ -0,0 +1,80 @@
+    const sourceKey: string | undefined = useMemo(
+        () => mappingArray[rowIndex]?.source,
+        [mappingArray, rowIndex],
+    );
```
You don't need to memoize strings

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/components/changes/TargetCell.tsx:
```diff
@@ -0,0 +1,80 @@
+
+    const sourceType = useMemo(() => {
+        return sourceOptions.find(option => option.value === sourceKey)?.type;
+    }, [sourceKey, sourceOptions]);
```
same here, all your types are either string or number

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/config/changes.tsx:
```diff
@@ -0,0 +1,176 @@
+): Array<Column> => {
+    const { formatMessage }: { formatMessage: IntlFormatMessage } =
+        useSafeIntl();
+    const columns: Array<Column> = [
```
I wonder if memoizing `columns` wouldn't save a couple renders, but I'm not sure

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/config/changes.tsx:
```diff
@@ -0,0 +1,176 @@
+    }
+    return columns;
+};
+
```
I would split the file in 2 here

beygorghor replied:
i'll leave like that for now ...

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/config/changes.tsx:
```diff
@@ -0,0 +1,176 @@
+    return columns;
+};
+
+type Props = {
```
nitpick: I would say `Params` i.o. `Props` and save `Props` for components

quang-le commented on hat/assets/js/apps/Iaso/domains/workflows/config/changes.tsx:
```diff
@@ -0,0 +1,176 @@
+    targetOptions: ChangesOption[];
+    handleUpdate: (
+        // eslint-disable-next-line no-unused-vars
+        key: 'target' | 'source',
```
keyof Mapping

quang-le reviewed (approved):
Nothing blocking, but a few nitpicks, and some stuff that should maybe go in another ticket:

- Follow-up: maybe we should add a button "restore order" for when users made changes, did not save and want to restore the table to its initial (active) order
- Follow-up: we should block saving when user has added a form but no condition. Back-end already returns a 400: 

https://user-images.githubusercontent.com/38907762/215439109-b21cee1c-b381-484a-860e-e14423775d5e.mov

- Changes : I would replace "Target form" with "Reference forms" to be consistent:
![Screenshot 2023-01-30 at 10 19 32](https://user-images.githubusercontent.com/38907762/215439261-9939d106-dc33-4a81-9157-6cdd650bfbf6.png)


Otherwise it works like a charm. The code is very legible. Nice work!

beygorghor commented:
> Nothing blocking, but a few nitpicks, and some stuff that should maybe go in another ticket:
> 
> * Follow-up: maybe we should add a button "restore order" for when users made changes, did not save and want to restore the table to its initial (active) order
> * Follow-up: we should block saving when user has added a form but no condition. Back-end already returns a 400:
 Regarding this, i would create two new tickets for that no ?

beygorghor committed:
Code review
